runtime/shared_heap.c

```diff
@@ -481,50 +497,118 @@ static intnat pool_sweep(struct caml_heap_state* local, pool** plist,
-    header_t* end = POOL_END(a);
-    mlsize_t wh = wsize_sizeclass[sz];
+    header_t* last_free_block = NULL;
+    const header_t* end = POOL_END(a);
+    const mlsize_t wh = wsize_sizeclass[sz];
     int all_used = 1;
     struct heap_stats* s = &local->stats;
 
-    /* conceptually, this is incremented by [wh] for every iteration
-       below, however we can hoist these increments knowing that [p ==
-       end] on exit from the loop (as asserted) */
+    a->next_obj = 0;
+
     work = end - p;
     do {
-      header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
-      if (hd == 0) {
-        /* already on freelist */
-        all_used = 0;
-      } else if (Has_status_hd(hd, caml_global_heap_state.GARBAGE)) {
+      header_t hd = (header_t)*p;
+
+      if( (char*)p + caml_plat_pagesize < (char*)end ) {
+        caml_prefetch((char*)p + caml_plat_pagesize);
+      }
+
+      /* The pools mark a block as being free by setting the tag to No_scan_tag
+        and the color to NOT_MARKABLE. The wosize is used to indicate the
+        number of contiguous free blocks that follow. The first field is a
+        pointer to the next free block beyond the immediately following
+        contiguous free blocks (if any). */
+
+      /* Check if the current block is garbage, if it is turn it into a free
+      block */
+      if (Has_status_hd(hd, caml_global_heap_state.GARBAGE)) {
         CAMLassert(Whsize_hd(hd) <= wh);
         if (Tag_hd (hd) == Custom_tag) {
           void (*final_fun)(value) = Custom_ops_val(Val_hp(p))->finalize;
           if (final_fun != NULL) final_fun(Val_hp(p));
         }
-        /* add to freelist */
-        atomic_store_relaxed((atomic_uintnat*)p, 0);
-        p[1] = (value)a->next_obj;
+
+        /* add to freelist. This could be optimised, we don't need
+        to write the free header if we're going to merge it with a prior
+        free block but it makes this codepath more complex. */
+        *p = POOL_FREE_HEADER(0);
+
         CAMLassert(Is_block((value)p));
 #ifdef DEBUG
         for (mlsize_t i = 1, wo = Wosize_whsize(wh); i < wo; i++) {
           Field(Val_hp(p), i) = Debug_free_major;
         }
 #endif
-        a->next_obj = (value*)p;
+
         all_used = 0;
         /* update stats */
         s->pool_live_blocks--;
         s->pool_live_words -= Whsize_hd(hd);
-        local->owner->swept_words += Whsize_hd(hd);
+        local->owner->swept_words += wh;
         s->pool_frag_words -= (wh - Whsize_hd(hd));
+
+        /* reload hd */
+        hd = POOL_FREE_HEADER(0);
+      }
+
+      /* If the current block was garbage (and is now a free block) or was
+      initially a free block, see if we can merge it with the last free block we
+      encountered or if we cannot then update the pointer in the last free block
+      to point to this one */
+      if (POOL_BLOCK_FREE_HD(hd)) {
+        /* if any block is free then this is no longer a full pool */
+        all_used = 0;
```

Comment:
`all_used` is cleared on line 543.
```suggestion
        CAMLassert (all_used == 0);
```

Reply:
I think this assertion is wrong. It's possible to have a pool where a free block (that was freed in the last sweep or before) is encountered before a garbage block that transitions to free (and sets all_used = 0). Reverting in https://github.com/ocaml/ocaml/pull/13616/commits/9ee312af447a65295414381edc08c6c4dba56815